[TRL-206] refactor: 여행 수정 기능 리팩터링

ttasjwi opened the pull request:
# JIRA 티켓
- [TRL-206]

---

# 작업 내역
- [x] 기간 수정 시 메서드 한번만 호출하도록 변경
- [x] TripPeriod가 기간에 관한 정보를 책임지는데 이에 관한 책임을 제대로 지고 있지 않는 문제 수정
- [x] TripPeriod 도메인 규칙에서 제약조건 추가
---


[TRL-206]: https://cosain.atlassian.net/browse/TRL-206?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

ttasjwi committed:
[TRL-206] refactor: 기간 변경 시 updateTripPeriod 메서드만 호출하도록 변경

ttasjwi committed:
[TRL-206] refactor: 시작날짜, 끝날짜 모두 null인 TripPeriod 생성 규칙 추가

- 시작날짜, 끝날짜 모두 null인 TripPeriod를 정의
- TripPeriod

ttasjwi committed:
[TRL-206] refactor: 여행의 소유자가 아닌 사람이 여행을 수정하려 할 때에 대한 방어로직 추가

ttasjwi committed:
[TRL-206] refactor: TripPeriod 생성 시 도메인 규칙에 맞게 방어로직 추가

ttasjwi committed:
[TRL-206] test: `TripPeriod.empty()` 테스트 추가

ttasjwi committed:
[TRL-206] feat: TripPeriod 간의 겹치는 구간을 구하는 메서드 작성 및 테스트

ttasjwi committed:
[TRL-206] feat: TripPeriod에 특정 date 포함 여부를 판단하는 메서드 추가

- TripPeriod의 contains 메서드를 통해 특정 date(LocalDate)가 해당 기간에 포함되는 지 여부를 판단한다.

ttasjwi committed:
[TRL-206] feat: Day가 TripPeriod에 속하는 지 여부를 판단하는 메서드 작성 및 테스트

ttasjwi committed:
[TRL-206] test: TripFixture에서 Trip create 시 EmptyPeriod를 기간으로 설정하도록 함

ttasjwi committed:
[TRL-206] feat: TripPeriod의 기간에 해당되는 날짜들을 스트림으로 반환하는 기능 추가

ttasjwi committed:
[TRL-206] feat: 기간이 정해진 여행의 기간을 빈 기간으로 바꿀 때의 예외 정의

ttasjwi committed:
[TRL-206] refactor: 여행의 기간 수정 시 TripPeriod와 협력하도록 리팩토링

ttasjwi committed:
[TRL-206] refactor: 기간에 변경사항이 없을 경우 리포지토리를 호출하지 않도록 함

ttasjwi committed:
[TRL-206] refactor: Trip을 가져올 때 day들을 한번에 가져오지 않는 문제 수정

- TripUpdateService에서 findByIdWithDays 메서드를 호출하지 않고 있어서, 이를 호출하도록 변경했다.

ttasjwi committed:
[TRL-206] refactor: TripUpdateCommand에서 날짜들 대신 TripPeriod를 가지도록 변경

- 표현계층에서 사용되는 TripUpdateDto에서는 순수하게 여행의 시작날짜, 끝 날짜를 받는다.
- 우리 애플리케이션에서는 TripPeriod 객체를 통해 '시작일부터 끝날짜까지의 연속된 기간'이라는 개념을 처리하면 여행 기간에 시작, 끝 날짜에 대한 검증을 TripPeriod가 책임지고 맡아 처리할 수 있기 때문에 더 응집도가 높은 코드가 될 것이라는 결론을 내렸다.
- 이후 애플리케이션 계층, 도메인 계층으로 전달할 때는 단순히 여행의 시작/끝날짜로 전달하기보다 TripPeriod로 변환을 거쳐서 전달하면 별도로 해당 계층에서 시작일,끝날짜에 대한 검증을 하지 않아도 된다.
- TripUpdateRequest에서 TripUpdateCommand로 변환하는 과정에서 startDate, endDate를 TripPeriod로 변환하여 전달함으로서 애플리케이션 레이어에 진입하기 전부터 우리 도메인 규칙의 제약을 걸고 잘못된 요청을 차단할 수 있다.

ttasjwi committed:
[TRL-206] chore: Trip의 기간 수정 메서드명 변경 및 TripUpdateService 내부 구현 메서드 배치 수정

- Trip.updatePeriod -> Trip.changePeriod
- TripUpdateService 메서드 구현 순서 변경

ttasjwi committed:
[TRL-206] test: 여행 기간 수정 기능 테스트 작성 및 일부 오류 수정

- 기존에 작성된 getNotOverlappedDays는 구현으로 은닉됐고 더 이상 필요 없으므로 테스트 코드를 지웠다.
- 기간 변경에 관해서는 Trip의 changePeriod 메서드만 퍼블릭 인터페이스로 노출됐다. 이 메서드만 테스트한다.
- 그런데 테스트 과정에서 여행 기간 변경 시 Trip의 tripPeriod가 새로운 tripPeriod으로 변경되지 않는 문제를 발견했고, 기간 변경 코드가 누락됐다는 것을 확인했다. 따라서 기간을 실제로 변경하는 코드를 추가했다.

ttasjwi committed:
[TRL-206] refactor: Trip의 changePeriod 메서드 내부 구현 수정

- deleteUnnecessaryDays : 불필요한 기간을 삭제 후 삭제된 기간 반환
- addNewDays : 새로운 기간을 생성 후 생성된 기간 반환

ttasjwi committed:
[TRL-206] test: 존재하지 않는 여행 수정 시도 시 예외 발생하는 케이스의 테스트 추가

ttasjwi commented on src/test/resources/application.yml:
```diff
@@ -72,6 +72,7 @@ spring:
         # show_sql: true # sout으로 하이버네이트 실행 SQL을 남기는데 이는 아래의 spring.logging.level.org.hibernate.SQL 옵션에서 log로 대체
         format_sql: true # sql을 로깅할 때 예쁘게 보여줌
         dialect: org.hibernate.dialect.MySQL8Dialect
+        create_empty_composites.enabled: true
```
TripPeriod의 모든 필드가 null일 경우, 하이버네이트가 TripPeriod 자체를 null로 가져오는 것이 기본 설정입니다.
따라서 필드가 모두 null인 TripPeriod로 가져오도록, 설정을 변경했습니다. 이는 배포에도 반영할 예정입니다.

ttasjwi commented on src/main/java/com/cosain/trilo/trip/command/application/service/TripUpdateService.java:
```diff
@@ -25,16 +23,10 @@ public class TripUpdateService implements TripUpdateUseCase {
 
-        List<Day> addDays = trip.updatePeriod(updateCommand.getStartDate(), updateCommand.getEndDate());
-        dayRepository.saveAll(addDays);
+        ChangeTripPeriodResult result = trip.updatePeriod(updateCommand.getStartDate(), updateCommand.getEndDate());
```
기존 코드는 서비스 계층에서, 겹치는 날짜를 Trip에게 물어보는 과정, 해당 날짜를 Trip에게 지워달라고 요청하는 과정, 그리고 Trip에 update 요청을 하여 생성되는 날짜를 얻어오는 과정이 나뉘어 있었습니다. 그리고 각 순간마다 Trip 객체의 정합성이 맞지 않는 경우가 발생했습니다.

저는 그래서 여기서 updatePeriod 명령 한 번으로, 불필요한 Day들을 지우고, 새로 생성되는 Day들을 추가하고 이들(생성되는 기간, 삭제되는 기간)을 반환하도록 하였습니다.

Java 문법에서는 하나의 메서드 결과로 하나의 객체 또는 값만을 반환하는 것이 허용되기 때문에 별도의 도메인 Dto인 ChangeTripPeriodResult라는 객체를 통해 이 두 결과를 담아 반환하도록 하였습니다. 이게 약간 부자연스럽게 느껴지긴 하지만, 차라리 명령을 두세번 거쳐서 진행하고 불필요한 내부 구현이 서비스 계층까지 노출되는 것보다는 낫다는 생각에 이 방식을 선택했습니다.

pia2011 replied:
`불필요한 내부 구현이 서비스 계층까지 노출되는 것` 동의합니다. 

여행 기간을 수정하게 되면 

- 기존 여행 기간에 속해있지만 새로 설정된 여행 기간에 속해있지 않는 Day 들을 삭제
- 기존 여행 기간에 속해있지 않고 새로 설정된 여행 기간에 속해있는 Day 들을 추가 

이렇게 두 가지를 처리해야 하는데, 저 또한 이는 여행 기간 수정으로 인해 발생하는 내부 구현에 속하는 것이라고 생각합니다. 굳!

ttasjwi commented on src/main/java/com/cosain/trilo/trip/command/domain/dto/ChangeTripPeriodResult.java:
```diff
@@ -0,0 +1,23 @@
+
+@Getter
+public class ChangeTripPeriodResult {
+
```
Trip의 TripPeriod를 변경함으로서, 삭제된 Day들과 생성된 Day들 각각을 리스트로 담기 위한 도메인 Dto 입니다.

ttasjwi commented on src/main/java/com/cosain/trilo/trip/command/domain/entity/Trip.java:
```diff
@@ -63,7 +64,10 @@ private Trip(Long id, Long tripperId, String title, TripStatus status, TripPerio
-        this.days = days;
+
+        if (days != null) {
+            this.days.addAll(days);
```
Trip이 가진 day들을 추가, 삭제하는 과정에서 days가 null인 경우가 발생하는 것을 확인하였습니다.
Builder를 통해 Trip을 생성시 day들을 전달하지 않으면 기본적으로 days에 null 이 할당되기 떄문입니다.

그래서 Trip의 객체 생성 시점부터 final 제어자를 통해 days에 빈 리스트를 할당하고 변하지 한 뒤, 이후에 생성자를 통해 days가 전달되면 이들을 리스트에 전부 addAll 하는 식으로 처리하기로 했습니다.

ttasjwi commented on src/main/java/com/cosain/trilo/trip/command/domain/entity/Trip.java:
```diff
@@ -74,54 +78,47 @@ public void changeTitle(String newTitle) {
      */
-    public List<Day> updatePeriod(LocalDate startDate, LocalDate endDate){
+    public ChangeTripPeriodResult updatePeriod(LocalDate startDate, LocalDate endDate){
+        List<Day> overlappedDays = getNotOverlappedDays(startDate, endDate);
```
기존의 updatePeriod 내부에 getNotOverlappedDays를 옮기고 updatePeriod 한번으로 모든 과정이 진행되도록 변경했습니다.

물론 getNotOverlappedDays 메서드는 이후에 단계적으로 리팩터링을 거치면서, 구현으로 은닉될 것입니다.

현재는 getNotOverlappedDays에 대한 테스트가 작성되어 있기 때문에 이 메서드는 public으로 열어뒀습니다.

ttasjwi commented on src/test/java/com/cosain/trilo/unit/trip/command/application/service/TripUpdateServiceTest.java:
```diff
@@ -45,15 +44,14 @@ public class TripUpdateServiceTest {
-        verify(tripRepository, times(1)).save(trip);
-        verify(dayRepository, times(0)).deleteDays(anyList());
-        verify(dayRepository, times(1)).saveAll(anyList());
+
```
기존에 리포지토리가 호출되는 횟수가 Undecided 상태일 때와 Decided 상태일 때 서로 달랐는데 이번에 코드를 바꾸면서 상태에 무관하게 모두 리포지토리가 호출되도록 되버렸습니다. 일단 이후에 수정하기로 합니다.

ttasjwi replied:
이는 이후의 커밋에서 수정됐습니다.

ttasjwi commented on src/main/java/com/cosain/trilo/trip/command/application/service/TripUpdateService.java:
```diff
@@ -22,17 +25,30 @@ public class TripUpdateService implements TripUpdateUseCase {
                 .orElseThrow(() -> new TripNotFoundException("일치하는 식별자의 Trip을 찾을 수 없음"));
     }
 
+    private void deleteUnnecessaryDays(List<Day> deletedDays) {
```
다시, Trip의 기간에 변경이 없을 경우 리포지토리를 호출하지 않도록 수정했습니다.

ttasjwi commented on src/test/java/com/cosain/trilo/unit/trip/command/application/service/TripUpdateServiceTest.java:
```diff
@@ -41,13 +42,14 @@ public class TripUpdateServiceTest {
-        verify(tripRepository).findById(anyLong());
-        verify(dayRepository).deleteDays(anyList());
-        verify(dayRepository).saveAll(anyList());
+        verify(tripRepository, times(1)).findById(anyLong());
```
기간 변경이 없을 시 리포지토리가 호출되지 않도록 되돌려서, 호출된 횟수를 다시 판별하도록 수정합니다.

ttasjwi commented on src/main/java/com/cosain/trilo/trip/command/application/service/TripUpdateService.java:
```diff
@@ -34,7 +34,7 @@ public void updateTrip(Long tripId, Long tripperId, TripUpdateCommand updateComm
 
     private Trip findTrip(Long tripId) {
-        return tripRepository.findById(tripId)
+        return tripRepository.findByIdWithDays(tripId)
```
전에 findByIdWithDays를 만들어두고, 서비스 계층에서 호출하지 않고 있더라구요. 그래서 수정했습니다.

pia2011 replied:
확인했습니다!

ttasjwi commented on src/main/java/com/cosain/trilo/trip/command/application/service/TripUpdateService.java:
```diff
@@ -33,6 +35,12 @@ public void updateTrip(Long tripId, Long tripperId, TripUpdateCommand updateComm
     }
 
+    private void validateTripUpdateAuthority(Trip trip, Long tripperId) {
+        if (!trip.getTripperId().equals(tripperId)) {
```
Trip의 소유자가 아닌 사람이 Trip을 수정할 수도 있어서 이를 일단 애플리케이션 계층인 TripUpdateService에서 막도록 했습니다.

pia2011 replied:
확인했습니다~!

ttasjwi commented on src/main/java/com/cosain/trilo/trip/command/domain/vo/TripPeriod.java:
```diff
@@ -22,11 +23,15 @@ public class TripPeriod {
     private LocalDate endDate;
 
     public static TripPeriod of(LocalDate startDate, LocalDate endDate) {
-        // 둘 다 null이 들어왔을 때
```
단순히 LocalDate startDate, LocalDate endDate 두개의 값으로 분리되어 있던 날짜 두개를
TripPeriod라는 객체 하나로 통합하고 시작일부터 끝까지라는 의미를 부여하였습니다.

따라서 이에 맞게 TripPeriod에 대한 규칙을 정의했습니다.
- startDate, endDate 둘다 null 이면 빈 기간으로 설정
- startDate, endDate 어느 한쪽만 Null 이면 예외 발생
- startDate는 endDate와 같거나 그 이전이여야 한다.

ttasjwi commented on src/main/java/com/cosain/trilo/trip/command/domain/vo/TripPeriod.java:
```diff
@@ -28,4 +34,11 @@ private TripPeriod(LocalDate startDate, LocalDate endDate) {
+     * 기간이 없는 TripPeriod를 반환합니다.
+     */
+    public static TripPeriod empty() {
+        return EMPTY_PERIOD;
```
- 시작일, 종료일 모두 null 인 경우는 도메인 규칙적으로 의미가 있는 경우(기간이 정해지지 않은 상태에 해당하는 기간)라서 별도로 처리하기로 했습니다.
- 본래 TripPeriod.empty는 싱글톤 인스턴스로 취급하고 싶었지만, JPA가 값객체 생성시 프로퍼티 주입을 통해 값 객체를 생성하므로 완전한 싱글톤은 지키지 못 하는 한계가 존재합니다. (JPA 엔티티와 도메인 엔티티를 분리한다면 이 문제를 해결할 수 있긴 합니다.) 그래서 이후 empty인지를 확인할 때는 `==`으로 비교할 수 없고, `equals`를 통해 동등성 관점에서 비교해야합니다.

ttasjwi commented on src/main/java/com/cosain/trilo/trip/command/domain/vo/TripPeriod.java:
```diff
@@ -46,4 +46,26 @@ private TripPeriod(LocalDate startDate, LocalDate endDate) {
+    public TripPeriod intersection(TripPeriod other) {
+        return hasNoIntersection(other)
+                ? empty()
+                : calculateIntersectionPeriod(other);
```
TripPeriod와 TripPeriod 간에 겹치는 기간에 해당하는 TripPeriod를 구하는 intersection 메서드를 작성했습니다.
메서드 명칭은 '교칩합'에서 차용했습니다. 하지만 다른 도메인적인 좋은 의미를 뜻하는 명칭이 있다면 바꿀 계획도 있습니다.

ttasjwi commented on src/main/java/com/cosain/trilo/trip/command/application/command/TripUpdateCommand.java:
```diff
@@ -1,23 +1,20 @@
     }
 
+    private TripUpdateCommand(String title, TripPeriod tripPeriod) {
+        this.title = title;
```
우리의 TripUpdateService는 입력 파라미터로 날짜 두개를 받았었습니다. startDate, endDate 두개의 인자를 말이죠.
하지만 단순히 변수몀만 startDate, endDate 였지 이것이 도메인적으로 의미 있는 입력은 아니였습니다. 사용자가 startDate, endDate를 순서를 뒤바꿔 전달하거나 하나만 null 인 경우를 전달하면 이것이 그대로 도메인 계층이나 애플리케이션 계층까지 전달되고 거기서 별도의 방어로직을 작성하게 되는 문제를 야기합니다.

TripPeriod를 받게 수정하면 TripUpdateCommand를 생성하는 시점부터 TripPeriod를 인자로 전달함을 강제할 수 있습니다. 앞에서 우리는 TripPeriod의 생성 규칙을 상세하게 기술했습니다. TripUpdateDto를 TripUpdateCommand로 변환하는 과정에서 날짜들을 TirpPeriod 생성에 전달함이 강제될 것입니다. 그리고 여기서 검증이 자연스럽게 도메인 규칙에 맞게 이루어지겠죠.

이를 통해 표현계층에서, 사용자 입력을 애플리케이션 계층의 도메인 용어로 변경하는 과정에서 유효성에 맞지 않으면 예외를 발생시킬 수 있습니다.

pia2011 replied:
확인했습니다. 

하지만 저는 TripPeriod 가 도메인 지식을 포함하는 값 객체이기 때문에 DTO에서 도메인 지식을 담는 다는게 조금 걸리기는 합니다. DTO의 목적이 데이터 전송이니까요. 이렇게 되면 결국 도메인 밖에서 도메인 지식에 의존하게 되는게 되고, 변경이 발생할 때 변경 지점이 늘어나게 됩니다.

즉, 결국에는 결합도가 증가하게 되는 것이기에 유지보수와 확장성 측면에서 고려했을 때 이것이 좋은 방법일지 잘 모르겠습니다.

ttasjwi replied:
고민해보니 TripUpdateCommand는 애플리케이션 계층에서 사용하는 명령 객체이므로 TripPeriod에 대해 알고 있어도 되지만, 이를 생성하는 과정에 TripPeriod가 노출될 필요도 없으므로 생성 로직을 수정해서 TripPeriod를 직접 생성 인자로 받지 않게 하겠습니다.

TripUpdateRequest에서 TripUpdateCommand로 변환하는 과정에서 TripPeriod 를 생성해서 전달하는데 이 생성 부분을 TripUpdateCommand에 은닉시키겠습니다.

ttasjwi commented on src/main/java/com/cosain/trilo/trip/command/domain/vo/TripPeriod.java:
```diff
@@ -68,4 +69,8 @@ private TripPeriod calculateIntersectionPeriod(TripPeriod other) {
     }
 
+    public boolean contains(LocalDate date) {
+        return !this.equals(empty()) &&
```
TripPeriod 기간에 특정 date가 포함되는 지 여부를 판단하는 메서드 `contains`를 작성했습니다.

ttasjwi commented on src/main/java/com/cosain/trilo/trip/command/domain/entity/Day.java:
```diff
@@ -39,4 +40,12 @@ private Day(LocalDate date, Trip trip) {
+     * @return 소속 여부
+     */
+    public boolean isIn(TripPeriod tripPeriod) {
+        return tripPeriod.contains(tripDate);
```
Day 자신이 특정 TripPeriod에 속하는 지 여부를 판단하는 로직을 작성했습니다. 이 과정에서 tripPeriod와 협력합니다.

ttasjwi commented on src/test/java/com/cosain/trilo/fixture/TripFixture.java:
```diff
@@ -28,6 +28,7 @@ public Trip createUndecided(Long id, Long tripperId, String title){
                 .tripperId(tripperId)
                 .title(title)
                 .status(this.status)
+                .tripPeriod(TripPeriod.empty())
```
테스트코드에서 사용하는 TripFixture 생성 시, 편의상 빌더패턴을 사용하는데 TripPeriod를 전달하지 않음으로 인해 TripPeriod가 null이 되는 문제가 발생합니다. 그래서 이를 수정했습니다.

ttasjwi commented on src/main/java/com/cosain/trilo/trip/command/domain/vo/TripPeriod.java:
```diff
@@ -73,4 +75,17 @@ public boolean contains(LocalDate date) {
+     * @return
+     */
+    public Stream<LocalDate> dateStream() {
+        if (this.equals(EMPTY_PERIOD)) {
```
TripPeriod에 속하는 연속적인 LocalDate들을 생성할 목적으로 dateStream 메서드를 작성했습니다.
List로 전달할 수도 있긴 하지만 Stream으로 직접 전달하는 것이 편리함이 많아서 이를 제공하기로 하였습니다.

ttasjwi commented on src/main/java/com/cosain/trilo/trip/command/domain/entity/Trip.java:
```diff
@@ -80,61 +86,45 @@ public void changeTitle(String newTitle) {
-                newDays.addAll(createDays(tripPeriod.getEndDate().plusDays(1), endDate));
-            }
+    public ChangeTripPeriodResult updatePeriod(TripPeriod newPeriod) {
+        // 기존이랑 기간이 같으면 변경 안 하고 반환
```
기존에는 TripPeriod를 변경할 떄 startDate, endDate만을 메서드 파라미터로 받았었습니다. 따라서 startDate, endDate가 잘못된 상황에 대해 검증하기 어려운 문제가 발생했습니다.

이제 메서드 파라미터로 TripPeriod를 전달함을 강제하면, 시작일-끝날짜에 대한 검증의 책임은 TripPeriod 생성시점에 위임되므로 더이상 updatePeriod 메서드에서 시작일-종료일의 유효성 판단 여부에 관심을 두지 않아도 됩니다.

그리고 여기서 전달된 TripPeriod와 Day, Trip이 협력하여 여행 수정기능을 구현하도록 하였습니다.

ttasjwi commented on src/main/java/com/cosain/trilo/trip/command/domain/entity/Trip.java:
```diff
@@ -80,61 +86,45 @@ public void changeTitle(String newTitle) {
-            currendDate = currendDate.plusDays(1);
-        }
+    public List<Day> getNotOverlappedDays(TripPeriod newPeriod) {
+        TripPeriod overlappedPeriod = tripPeriod.intersection(newPeriod);
```
삭제되는 Day를 구할 때는, 기존 기간과 새로운 기간의 intersection(교집합)을 구하고 이에 속하지 않는 기존 Day들을 제거해주면 됩니다.

ttasjwi commented on src/main/java/com/cosain/trilo/trip/command/domain/entity/Trip.java:
```diff
@@ -80,61 +86,45 @@ public void changeTitle(String newTitle) {
-        return days;
+    private List<Day> getCreateDays(TripPeriod newPeriod) {
+        return newPeriod.dateStream()
+                .filter(date -> !tripPeriod.contains(date))
```
생성되는 기간을 구할 때는 새로운 기간의 date들을 반복 순회하면서, 기존 구간에 포함되지 않는 날짜들에 한해서만 Day 객체를 생성하도록 하면 됩니다. 이 과정에서도 TripPeriod의 적극적인 참여가 필요합니다.

ttasjwi commented on src/main/java/com/cosain/trilo/trip/command/domain/entity/Trip.java:
```diff
@@ -105,24 +107,27 @@ public ChangeTripPeriodResult changePeriod(TripPeriod newPeriod) {
             status = TripStatus.DECIDED;
         }
-        List<Day> deleteDays = getNotOverlappedDays(newPeriod);
+        this.tripPeriod = newPeriod;
```
테스트를 작성해서 테스트를 돌리다보니, 테스트가 깨지는 것을 확인했는데
새로운 기간으로 바꾸는 코드가 누락되어있더군요. 그래서 수정했습니다.

테스트의 중요성을 다시 확인합니다.

ttasjwi commented on src/test/java/com/cosain/trilo/unit/trip/command/domain/entity/TripTest.java:
```diff
@@ -70,37 +72,480 @@ public void trip_has_changed_title() {
-    @DisplayName("'getNotOverlappedDays 로 겹치지 않는 날짜를 가져올 때")
-    class When_getNotOverlappedDays {
+    @DisplayName("ChangePeriod 테스트")
+    class ChangePeriodTest {
```
changePeriod 메서드의 테스트가 중요할 것 같아서 이 부분에 여러가지 케이스를 나누고 시간을 많이 투자했습니다.(노가다...)

![image](https://user-images.githubusercontent.com/88282356/234585688-a2924147-2d1a-4729-946d-dc167987fb96.png)

ttasjwi committed:
[TRL-206] refactor: TripUpdateCommand에 TripPeriod 전달하던 것을 은닉

- 표현 계층(TripUpdateRequest)에서 TripUpdateCommand를 생성할 때 생성 인자로 TripPeriod를 전달하는 상황이 발생했었음. 이는 도메인 규칙에 대해서 표현 계층이 많은 것을 알게 됨을 의미한다.
- TripUpdateCommand를 생성할 때 startDate, endDate만을 전달하고, 이를 기반으로 TripPeriod 를 생성하는 책임은 TripUpdateCommand에게 위임한다. 이렇게하면 TripUpdateRequest는 도메인 계층의 TripPeriod를 몰라도 된다.
- 서비스 계층에서는 기존 코드를 변경할 필요는 없다. 기존대로 TripUpdateCommand로부터 변경할 TripPeriod를 그대로 꺼내오고, 사용하기만 하면 된다.

ttasjwi committed:
[TRL-206] fix: TripRepository.findByIdWithDays 의 JPQL 수정

- Inner Join을 통해 가져오기 때문에 Day가 존재하지 않으면 Trip을 가져오지 않는 문제가 발생했다.
- LEFT JOIN을 통해 Day의 존재 여부에 관계 없이 Trip을 반드시 가져오도록 수정했다. Day가 존재하면 Day들도 같이 가져온다.
- 이 상황에 대한 테스트 코드를 추가적으로 작성했다.

pia2011 commented on src/main/java/com/cosain/trilo/trip/command/application/command/TripUpdateCommand.java:
```diff
@@ -3,14 +3,16 @@
-    public static TripUpdateCommand of(String title, TripPeriod tripPeriod) {
-        return new TripUpdateCommand(title, tripPeriod);
+    public static TripUpdateCommand of(String title, LocalDate startDate, LocalDate endDate) {
+        return new TripUpdateCommand(title, TripPeriod.of(startDate, endDate));
```
확인했습니다! 이제 도메인 지식을 표현 계층에서 분리시켜 응집도를 높였네요.

pia2011 commented on src/main/java/com/cosain/trilo/trip/command/domain/repository/TripRepository.java:
```diff
@@ -10,7 +10,7 @@
 
     @Query("SELECT t" +
-            " FROM Trip as t JOIN FETCH t.days" +
+            " FROM Trip as t LEFT JOIN FETCH t.days" +
```
Day 가 존재하지 않을 수 있으므로 `LEFT JOIN` 으로 확인했습니다.
